File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    def findCircleNumMineNotWorking(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        parents = {}
        for i in range(n):
            parents[i]=i


        for i in range(n):
            for j in range(i+1,n):
                if isConnected[i][j]:
                    temp = parents[i]
                    for k in range(len(parents)):
                        if parents[k] == temp:
                            parents[k] = j
                    print(parents)

        return len(set(parents.values()))

    def findCircleNum(self, M: List[List[int]]) -> int:
       
        if not M:
            return 0
        
        n = len(M)
        visit = [False]*n
        
        def dfs(u):
            for v in range(n):
                if M[u][v] ==1 and visit[v] == False:
                    visit[v] = True
                    dfs(v)
        
        count = 0
        for i in range(n):
            if visit[i] == False:
                count += 1
                visit[i] = True
                dfs(i)
        
        return count

File: 0547-number-of-provinces/0547-number-of-provinces.c
void dfs(int i, int** isConnected, int len, int* visit) {
    for(int j=0; j<len; ++j) {
        if(isConnected[i][j] && !visit[j]) {
            visit[j] = 1;
            dfs(j, isConnected, len, visit);
        }
    }
}
int findCircleNum(int** isConnected, int isConnectedSize, int* isConnectedColSize) {
    int count = 0;
    int* visit = (int*)calloc(isConnectedSize, sizeof(int));
    for(int i=0; i<isConnectedSize; i++)
    {
        if(!visit[i])
        {
            count++;
            visit[i]=1;
            dfs(i, isConnected, isConnectedSize, visit);
        }
    }
    return count;  
}